### PR TITLE
ENH: GH17054: read_html() handles rowspan/colspan and infers headers

### LINKED ISSUE #17054: ENH: read_html to handle rowspan, colspan
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
pd.read_html('https://www.ssa.gov/policy/docs/statcomps/supplement/2015/5h.html')[0]
```

This has complex table headings:

annual_statistical_supplement__2015_-_beneficiary_families_with_oasdi_benefits_in_current-payment_status__5_h_

`read_html` output begins with:

```
                                                 Year                            Retired-worker families        Survivor families             Disabled-worker families                                Unnamed: 4_level_0                              Unnamed: 5_level_0 Unnamed: 6_level_0 Unnamed: 7_level_0 Unnamed: 8_level_0 Unnamed: 9_level_0 Unnamed: 10_level_0 Unnamed: 11_level_0  \
                                          Worker only                                  Worker and wife?a  Non-disabled widow only        Widowed mother or father and?                                       Worker only                            Worker, wife,?b and?  Worker and spouse Unnamed: 7_level_1 Unnamed: 8_level_1 Unnamed: 9_level_1 Unnamed: 10_level_1 Unnamed: 11_level_1
                                                  All                                                Men                    Women                              1?child                                        2?children                              3 or more children                All                Men              Women            1?child  2 or more children Unnamed: 11_level_2
0                                                 NaN                                 Number?(thousands)                      NaN                                  NaN                                               NaN                                             NaN                NaN                NaN                NaN                NaN                 NaN                 NaN
1                                                1945                                                416                      338                                   78                                               181                                              95              86.00              48.00              24.00              .?.?.               .?.?.               .?.?.
```

(row 0 of the output is probably something one would have to manually eliminate)
#### Problem description

For HTML headings with rowspan and colspan elements, `read_html` has undesirable behavior. Basically `read_html` packs all heading `<th>` elements in any particular row to the left, so any particular column no longer has any association with the `<th>` elements that are actually above it in the HTML table.

Ample discussion here about the analogous pandas+Excel test case: https://github.com/pandas-dev/pandas/issues/4679

Relevant web discussions:
- https://stackoverflow.com/questions/23703638/bug-in-pandas-read-html-method
- https://stackoverflow.com/questions/28763891/what-should-i-do-when-tr-has-rowspan
- https://johnricco.github.io/2017/04/04/python-html/   (using his example)

This may be an issue with the underlying parsers and cannot be solved well in pandas. This appears to be the behavior with both lxml and bs4/html5lib.

#### Expected Output

Each column should be associated with the `<th>` elements above it in the table. This might be a multi-row column name (as it is now) (a `MultiIndex`?) or a tuple (presumably if the argument `tupleize_cols` is set to `True`). Instead, currently, column _n_ is associated with the _n_ th `<th>` entry in the table row regardless of the settings of rowspan/colspan.

It may be this is possible to do properly in current pandas in which case I apologize for filing the issue (but I'd be happy to know how to do it).

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.13.final.0
python-bits: 64
OS: Darwin
OS-release: 16.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.US-ASCII
LOCALE: None.None

pandas: 0.20.3
pytest: None
pip: 9.0.1
setuptools: 36.2.0
Cython: 0.26
numpy: 1.13.1
scipy: 0.19.1
xarray: None
IPython: 5.3.0
sphinx: 1.6.3
patsy: None
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.2
feather: None
matplotlib: 2.0.2
openpyxl: 2.4.7
xlrd: 1.0.0
xlwt: None
xlsxwriter: None
lxml: 3.7.3
bs4: 4.5.3
html5lib: 1.0b10
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by jowens**
- [x] closes #17054
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master -u -- "*.py" | flake8 --diff``
 - [x] whatsnew entry

### CODE REVIEW

**Comment by jowens** [comment]
@gfyoung can you help me reconcile failing tests in the CI checks vs. clean tests on my machine?

```
$ py.test-2.7 pandas/tests/io/test_html.py
============================= test session starts ==============================
platform darwin -- Python 2.7.13, pytest-3.1.3, py-1.4.34, pluggy-0.4.0
rootdir: /Users/jowens/Documents/src/pandas-jowens, inifile: setup.cfg
collected 76 items

pandas/tests/io/test_html.py ............................................................................

========================== 76 passed in 49.13 seconds ==========================
```

**Comment by gfyoung** [comment]
@jowens : Judging from your Travis build output, here's what I see:

* You have a style-check error on the Python 2.7 machine.  However, all tests are passing.  That indicates your patch is working for Python 2.7.  You should patch this first to get that build working at least.

* Your changes are failing for Python 3.x, meaning that they are perhaps not compatible with Python 3.  I would suggest that you create a separate environment using Python 3 as your interpreter (I see you are using Python 2), install your branch, and rerun tests to see if you can replicate the failures.

If you can, then continue developing in the Python 3.x environment to get tests passing.  If you can't replicate, then that's going to be tricky to debug, but we can address that *if* we get there.

**Comment by jowens** [comment]
Can you point me to the "style-check error"?

**Comment by gfyoung** [comment]
The error can be found <a href="https://travis-ci.org/pandas-dev/pandas/jobs/258193317#L1818">here</a>.  BTW, you can also see the error yourself by running the style-check command in the issue PR description:

~~~
git diff upstream/master -u -- "*.py" | flake8 --diff
~~~

**Comment by jowens** [comment]
Yeah, I did that before checking in:

```
$ git diff upstream/master -u -- "*.py" | flake8-2.7 --diff
$
```

Not sure what went wrong. I'm on it!

**Comment by jowens** [comment]
Cool! @gfyoung what's next?

**Comment by jowens** [comment]
@gfyoung any chance of moving this forward? I'd prefer not to have a private build of Pandas for the rest of my life.

**Comment by gfyoung** [comment]
@jowens : Sorry about that!  My inbox is very full most of the time, so I think I just missed your comment about a month ago.

Let's do this:
* Rebase onto `master` (and fix the merge conflict) to see if tests still pass
* Ping @TomAugspurger for further review of the PR

**Comment by jowens** [comment]
@gfyoung FWIW the merge conflict is only "whatsnew"

**Comment by jowens** [comment]
@gfyoung because I *really* don't want to get this wrong and have a commit that's gonna be enormous, can you sketch out how I "rebase onto master" properly?

**Comment by gfyoung** [comment]
@jowens: True, but it's been a month since you've run the tests on Travis, Appveyor, etc.  I just want to make sure that tests still pass.

**Comment by gfyoung** [comment]
> @gfyoung because I really don't want to get this wrong and have a commit that's gonna be enormous, can you sketch out how I "rebase onto master" properly?

I would recommend to try using GitHub's UI editor for resolving the merge conflict.  If that doesn't work out, I can give you further suggestions, but they will be all local (and probably) less visual.

**Comment by gfyoung** [comment]
Don't worry about rebasing onto master yet.  Let's make sure this PR is merge-able first.

**Comment by jowens** [comment]
@gfyoung OK, looks like it's running and conflict-free. Stay tuned!

**Comment by jowens** [comment]
@TomAugspurger can you help me take the next step with this PR? Thanks @gfyoung!

**Comment by TomAugspurger** [comment]
I won't have a chance to review this in detail for a while, so if you think
it's good then feel free to merge @gfyoung.

**Comment by jowens** [comment]
@TomAugspurger @gfyoung could either one of you guys please take a look?

File: doc/source/whatsnew/v0.21.0.txt
Comment (gfyoung) [inline doc/source/whatsnew/v0.21.0.txt]: Add these lines back.

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: As tree-like is not a Python data-type, we should re-specify this.

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: Let's take advantage of this time to re-specify this as a proper data type if possible.

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: string --> str

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: Just say "boolean indicating the object matches the tag `tag`"

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: string --> str

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: Let's try to re-specify as something other than node-like.

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: Use the same boolean description as above: "Boolean indicating whether..."

File: pandas/io/html.py
Comment (jowens) [reply pandas/io/html.py]: I totally copied "node-like" from other parts of the documentation. Suggestion of what I should use?

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: Let's take advantage of this time to re-specify this as a proper data type if possible.

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: Let's take advantage of this time to re-specify this as a proper data type if possible.

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: Say "tuple of (header, body, footer)"

File: pandas/io/html.py
Comment (gfyoung) [reply pandas/io/html.py]: @TomAugspurger : If you think "node-like" or "tree-like" is OK for this documentation, then we can ignore this comment.  However, IMO we should have proper data types in the doc-strings.

File: pandas/io/html.py
Comment (gfyoung) [reply pandas/io/html.py]: @jowens : See discussion <a href="https://github.com/pandas-dev/pandas/pull/17089#discussion_r137319069">above</a>.  I don't want to buck convention if this has been the thing to use.  However, the best thing you can do is try to determine the data-type yourself when you call the function.  For example, add a `print(type(obj))` in your implementation to see what it is.

File: doc/source/whatsnew/v0.21.0.txt
Comment (gfyoung) [reply doc/source/whatsnew/v0.21.0.txt]: Also, your `whatsnew` has a merge conflict, so you definitely need to fix that before we merge this.

File: pandas/io/html.py
Comment (gfyoung) [reply pandas/io/html.py]: Just provide the data-type that you get when you add `print(type(obj))` to the implementation of the function.  I'm not entirely sure what it is myself, hence why I suggested you do that to get an exact answer.

**Comment by jowens** [comment]
OK @gfyoung next step?

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: I think you still need to supply the actual data type of the input i.e. what is the output of type(`doc`) when used in this function?  Just add a line in the function that prints out the `type` and recompile `pandas` to include it.  Then see what it is and then add that to the documentation.

Comment (jowens) [reply pandas/io/html.py]: OK. It depends on what parser you use, though (BeautifulSoup or lxml). It's the BS4 or lxml parser object. I will check exact types.

Comment (gfyoung) [reply pandas/io/html.py]: Arghhh!  I wish you had told me this sooner. :cry:

Comment (gfyoung) [reply pandas/io/html.py]: For future reference, we are *generally* very picky about the data types in the doc-string.  However, as the data-type is dependent on an external factor (i.e. engine), we drop the data type and immediately dive into the description of the variable.

Comment (gfyoung) [reply pandas/io/html.py]: Here is what I mean by that for example:

BEFORE
~~~python
row : node-like
     A row in the HTML table
~~~

AFTER
~~~python
row : a row in the HTML table
~~~

Comment (jowens) [reply pandas/io/html.py]: Sorry. Didn't know that until I dived into the documentizing. So right now I am consistent about this kind of docstring:

```
        rows : list of HTML row elements
            A list of row elements of a table, usually <tr> or <th> elements.
```

but the `row element` here is also dependent on the parser (it's either a BS4 or lxml row element). This does seem like the right way to describe it though. (I left the documentation structured as is in the way it described subclassed functions; there is an virtual function, and that's what's documented, and then subclassed functions specific to the parser that are not documented.)

So: 

- Would you like me to back out the changes I made (the ones like the above) and redo them as you suggest?
- Based on what you suggest, `parser object` does seem like the right way to describe it ...?

Comment (gfyoung) [reply pandas/io/html.py]: @jowens : Instead of providing a "data-type" like `parser object`, go right into describing the data input, so in the example you provided above, do this:

~~~
rows : a list of row elements of a table, usually <tr> or <th> elements.
~~~

Comment (jowens) [reply pandas/io/html.py]: OK. For where I have "parser object", here's the actual type: `<class 'bs4.BeautifulSoup'>`. Suggestion for what to put instead of "parser object"?

Comment (gfyoung) [reply pandas/io/html.py]: Just do this:

~~~python
doc : the DOM from which to parse the table element.
~~~

Comment (jowens) [reply pandas/io/html.py]: Many thanks. Next round pushed. You are a very patient person.

**Comment by jowens** [comment]
Hit me with more changes, @gfyoung!

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: This is repetitive.  Just put your description in the line above.

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: Write:

~~~python
is_tag_equal : boolean
    boolean indicating if the object is equal to tag 'tag'
~~~

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: Visually, this formatting looks weird.  Can you indent any subsequent lines in your description and avoid lone words on a line.  For example:

~~~python
header : list of rows, each of which is
         a list of header elements
~~~

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: otherwise --> Otherwise
Add period at end.

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: Add newline above your comment.  Rewrite so that "does" is not alone in this line.

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: Similar to `is_tag_equal` above, provide a variable name here and then the data type.

File: pandas/tests/io/test_html.py
Comment (gfyoung) [inline pandas/tests/io/test_html.py]: Reference issue number above

Comment (gfyoung) [reply pandas/tests/io/test_html.py]: Do this for all of the tests that you added.

**Comment by jowens** [comment]
@gfyoung @TomAugspurger Next steps?

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: Let's move this onto the previous line with a colon after `ValueError`

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: Add a brief doc-string here.

**Comment by gfyoung** [comment]
@jowens : Sorry for not looking at this sooner.  I made some additional comments, but it's definitely well on its way to getting merged.  Feel free to ping us again as soon as you finish addressing my comments.

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: Remove the "if"

**Comment by jowens** [comment]
@gfyoung I expect no issues with the build since it's documentation-changes-only. FWIW these are documentation issues that predate any of my changes (but I don't mind cleaning them up).

**Comment by gfyoung** [comment]
@jowens : Understood.  I suggested them because those docs are showing up in your diff, so they're fair game as far as I'm concerned. :smile:

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: Nit: Add newline above this one.

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: Nit: add newline above this one.

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: Nit: add newline above this one.

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: Nit: add newline above this one.

File: pandas/io/html.py
Comment (gfyoung) [inline pandas/io/html.py]: Nit: add newline above this one.

**Review by gfyoung — APPROVED** [review]
LGTM.  Thanks for the changes!

**Comment by jowens** [comment]
@gfyoung Any idea why Travis CI failed here? It passed before and all I changed recently was documentation.

**Comment by gfyoung** [comment]
@jowens : This was handled in in #17588.  Rebase on `master` and push again.

**Comment by jowens** [comment]
@TomAugspurger Next steps?

**Comment by jowens** [comment]
@TomAugspurger @gfyoung What can I do to move this forward?

**Comment by gfyoung** [comment]
@TomAugspurger : Can you take a look at this?  I've already approved, and @jowens and I are awaiting your comments before proceeding.

File: pandas/io/html.py
Comment (TomAugspurger) [inline pandas/io/html.py]: Don't need the parens

File: pandas/io/html.py
Comment (TomAugspurger) [inline pandas/io/html.py]: You can remove these blank lines between parameters

File: pandas/io/html.py
Comment (TomAugspurger) [inline pandas/io/html.py]: you can remove this line (multiple returns are always tuples)

File: pandas/io/html.py
Comment (TomAugspurger) [inline pandas/io/html.py]: What's the plan here? Do we have a GH issue open for replacing the old with the new? i.e., what does the person reading the code 3 years from now need to know whether this has been resolved?

File: pandas/io/html.py
Comment (TomAugspurger) [inline pandas/io/html.py]: This seems to be untested.

File: pandas/io/html.py
Comment (TomAugspurger) [inline pandas/io/html.py]: What happens with a table of all "headers". Does the condition need to be `while body_rows and all(...)` (is it possible to get to this situation, and do you have a test for it)?

File: pandas/io/html.py
Comment (TomAugspurger) [inline pandas/io/html.py]: I think this is equivalent to

```python
list(chain.from_iterable([text] * nc for text, nc in zip(cols_text, col_colspans)))
```

which I find a bit more readable. Can you confirm if that works? (import `chain` from itertools).

Comment (TomAugspurger) [reply pandas/io/html.py]: Hmm, reading a bit further, maybe not? But I think that's close. Could you post a sample of `col_colspans`, `col_rowspans`, and `cols_texts`? Are they all just flat lists of ints / strs?

File: pandas/io/html.py
Comment (TomAugspurger) [inline pandas/io/html.py]: Untested?

File: pandas/io/html.py
Comment (TomAugspurger) [inline pandas/io/html.py]: Untested?

File: pandas/io/html.py
Comment (TomAugspurger) [inline pandas/io/html.py]: This could be a `versionmodified`, though it may not be necessary.

**Review by TomAugspurger — COMMENTED** [review]
Overall things look pretty good. I don't completely understand the logic in `_expand_colspan_rowspan`, and there were several lines that seem to be untested (perhaps they're dead code now?).

Could you attempt to refactor those `lmaps` to use list-comprehensions and `chain`? Could you add some more tests for

- table with just a `thead`, but no body
- table with just a `tbody`, but each row is a `th`

And cases where `rowspan` / `colspan` vary for each of those.

**Comment by jreback** [comment]
needs to be rebased

**Comment by jowens** [comment]
@TomAugspurger made a lot of good suggestions that I simply haven't had time to get to yet. Is there an impending deadline here?

**Comment by jreback** [comment]
@jowens nope. just pinging on older pr's. rebase / fixup when you can.

**Comment by ostrokach** [comment]
@jowens I just tried out this branch for extracting tables from PubMed Central XML files, and and it works much much better than the current `read_html` implementation (which was basically unusable)! Thank you for your work guys! Hope this can make it into the master branch soon :).

**Comment by jreback** [comment]
can you ``git merge master`` resolve conflicts and push again (and move whatsnew to 0.23.0)

**Comment by jowens** [comment]
@TomAugspurger's suggestions are still all good ones, but I haven't had a full day where I can sit down and do them.

**Comment by wkerzendorf** [comment]
great work! can't wait for it to get merged

**Comment by jowens** [comment]
I'll repeat that "@TomAugspurger's suggestions are still all good ones, but I haven't had a full day where I can sit down and do them." :)

**Comment by TomAugspurger** [comment]
No rush, we're patient :)

**Comment by cancan101** [comment]
any updates on this PR?

**Comment by adamhooper** [comment]
@jowens Hi! This feature is at the top of my backlog this week. So effectively, _I_ have a full day where I can sit down and incorporate @TomAugspurger's suggestions. Would that be okay with you?

**Comment by jowens** [comment]
Yes PLEASE! I would be delighted (and so would the other folks in this thread).

**Comment by WillAyd** [comment]
Superseded by #21487

**Pull request closed without merging**
